src/coreclr/nativeaot/Common/src/Internal/Runtime/CompilerHelpers/StartupCodeHelpers.cs

```diff
@@ -90,7 +90,8 @@ internal static int GetLoadedModules(TypeManagerHandle[] outputModules)
 
         private static void AddModule(TypeManagerHandle newModuleHandle)
         {
-            if (s_modules == null || s_moduleCount >= s_modules.Length)
+            Debug.Assert(s_modules != null);
```

Comment:
If you notice, I remove `s_modules == null` check, since `AddModule` called in C# only when `s_modules != null` so I decide that better express my knowledge about that in `Debug.Assert`. I prefer to not add `!` at extra cost, if this is not add runtime penalty. That's why I litter code with `Debug.Assert` to state what I know about code. I think about that as conditional `!` where I express. I assume that if assert was triggered then somebody screw up somewhere, which is same as `NullReferenceException`.

All first commit have very strange cases. Second one is mostly ride back on my decision to use `Debug.Assert`.

> but I don't know if there's a [NotNullIfBlahBlah] that can express that.

These guys good for parameters, but from what I know they do not work on fields or combination of fields. So at least `Debug.Assert` not only silence compiler, but also express my brave nature trusting myself that some invariant holds.

Reply:
> That's why I litter code with Debug.Assert to state what I know about code. 

The `!' was designed precisely for that - "I know that this cannot be null here and that's why I'm putting this extra character". The C# designers could have chosen not to have the `!` because it's redundant with just adding `Debug.Assert` everywhere.

Reply:
That's actually an argument for me. The only thing which I like about `Debug.Assert` is that I can put rationale why it is not null. Not in this case for sure.